Source/WebKit/NetworkProcess/NetworkProcess.cpp

```diff
@@ -392,6 +396,31 @@ void NetworkProcess::createNetworkConnectionToWebProcess(ProcessIdentifier ident
         session->storageManager().startReceivingMessageFromConnection(connection.connection());
 }
 
+void NetworkProcess::addAllowedFirstPartyForCookies(WebCore::ProcessIdentifier processIdentifier, URL&& firstPartyForCookies)
+{
+    m_allowedFirstPartiesForCookies.ensure(processIdentifier, [] {
+        return HashSet<URL> { };
+    }).iterator->value.add(WTFMove(firstPartyForCookies));
+}
+
+bool NetworkProcess::allowsFirstPartyForCookies(WebCore::ProcessIdentifier processIdentifier, const URL& firstParty)
+{
+    if (!decltype(m_allowedFirstPartiesForCookies)::isValidKey(processIdentifier)) {
+        ASSERT_NOT_REACHED();
+        return false;
+    }
+    auto iterator = m_allowedFirstPartiesForCookies.find(processIdentifier);
+    if (iterator == m_allowedFirstPartiesForCookies.end()) {
+        ASSERT_NOT_REACHED();
+        return false;
+    }
+    if (!decltype(iterator->value)::isValidValue(firstParty))
+        return false;
+    auto result = iterator->value.contains(firstParty);
```

Comment:
Would it suffice to validate security origins instead of URLs?

Wouldn't the following be racy?
1. Do a fragment navigation
2. Access document.cookie

Step 1 will do a sync IPC to the UIProcess to check the policy decision which will send the IPC to the network process. One the process receives the policy decision, it asks for cookies from the network process. I don't think there is any guarantee the network process got the notification from the UIProcess about the new URL.

Also, right now, we have injected bundle SPI for the client to answer the navigation policy decision from its injected bundle. When the client accepts the navigation from its injected bundle, the UIProcess is not involved at all and the network process won't know about the new URL.